src/core/regimes.rkt

```diff
@@ -175,127 +175,131 @@
-      ;; The default, not making a new split-point, gets a bonus of min-weight
-      (let ([acost (fl- a-cost min-weight)])
-        (for ([prev-split-idx (in-range 0 point-idx)])
-          ;; For each previous split point, we need the best candidate to fill the new regime
-         (when 
-          (vector-ref can-split-vec (+ prev-split-idx 1))
-          (let ([best #f] [bcost #f])
-            (for ([cidx (in-naturals)] [psum (in-vector flvec-psums)])
-              (let ([cost (fl- (flvector-ref psum point-idx)
-                             (flvector-ref psum prev-split-idx))])
-                (when (or (not best) (fl< cost bcost))
-                  (set! bcost cost)
-                  (set! best cidx))))
-            (define temp (fl+ (flvector-ref v-alt-cost prev-split-idx) bcost))
-            (when 
-              (fl< temp acost)
-              (set! acost temp)
-              (set! a-cost acost)
-              (set! a-best best)
-              (set! a-prev-idx prev-split-idx)))))
-        (flvector-set! vec-alt-cost point-idx a-cost)
-        (vector-set! vec-cidx point-idx a-best)
-        (vector-set! vec-pidx point-idx a-prev-idx)))
-  (values vec-alt-cost vec-cidx vec-pidx))
+  (for ([point-idx (in-range 0 number-of-points)])
+   ;; Set and fill temporary vectors with starting data
+   ;; #f for best index and positive infinite for best cost
+   (vector-fill! best-alt-idx #f)
+   (vector-fill! best-alt-cost +inf.f)
 
-  ;; We get the initial set of cse's by, at every point-index,
-  ;; accumulating the candidates that are the best we can do
-  ;; by using only one candidate to the left of that point.
-  (define (initial)
-    (define vec-acost (make-flvector num-points))
-    (define vec-cidx (make-vector num-points))
-    (define vec-pidx (make-vector num-points))
-    (define vec-temp (make-flvector num-candidates))
-    (for ([point-idx (in-range num-points)])
-      ;; record the cost from each candidate
-      (for ([cand-idx (range num-candidates)])
-       (flvector-set! vec-temp cand-idx
-        (flvector-ref (vector-ref flvec-psums cand-idx) point-idx)))
-      ;; Find the min, no built in function to find smallest fl in vector
-      (define min (flvector-ref vec-temp 0))
-      (define min-idx 0)
-      (for ([val vec-temp] [idx (range num-candidates)])
-        (cond [(< val min)
-               (set! min-idx idx)
-               (set! min val)]))
-      (flvector-set! vec-acost point-idx (fl min))
-      (vector-set! vec-cidx point-idx min-idx)
-      (vector-set! vec-pidx point-idx num-points))
-    (values vec-acost vec-cidx vec-pidx))
+   ;; For each alt loop over its vector of errors
+   (for ([alt-idx (in-naturals)] [alt-error-sums (in-vector flvec-psums)])
+    ;; Loop over the points up to our current point
+    (for ([prev-split-idx (in-range 0 point-idx)])
+     ;; Check if we can add a split point
+     (when (vector-ref can-split-vec (+ prev-split-idx 1))
```

Comment:
This `when` can be moved into a `#:when`

Reply:
Also I think instead of `vector-ref` we can just also iterate over `can-split-vec` in parallel